server/pulp/server/async/app.py

```diff
@@ -87,14 +84,31 @@ def initialize_worker(sender, instance, **kwargs):
         get_resource_manager_lock(sender)
 
 
+@worker_shutdown.connect
+def shutdown_worker(signal, sender):
+    """
+    Called when a worker is shutdown.
+
+    So far, this just cleans up the database by removing the worker's record in
+    the workers collection.
+
+    :param signal:   The signal being sent to the worker
+    :type  signal:   int
+    :param instance: The hostname of the worker
+    :type  instance: celery.apps.worker.Worker
+    """
+    tasks._delete_worker(sender.hostname, normal_shutdown=True)
+
+
 def get_resource_manager_lock(name):
     """
-    Tries to acquire the resource manager lock. If the lock cannot be acquired immediately, it
-    will wait until the currently active instance becomes unavailable, at which point the worker
-    cleanup routine will clear the lock for us to acquire. A worker record will be created so that
-    the waiting resource manager will appear in the Status API. We override the SIGTERM signal
-    handler so that that the worker record will be immediately cleaned up if the process is killed
-    while in this states.
+    Tries to acquire the resource manager lock.
+
+    If the lock cannot be acquired immediately, it will wait until the
+    currently active instance becomes unavailable, at which point the worker
+    cleanup routine will clear the lock for us to acquire. A worker record will
+    be created so that the waiting resource manager will appear in the Status
+    API.
```

Comment:
I think adding a line here that the worker record is expected to be cleaned up upon shutdown would be good. It can remain silent on where/how that happens.